edge-agent/src/Microsoft.Azure.Devices.Edge.Agent.Kubernetes/edgedeployment/EdgeDeploymentController.cs

```diff
@@ -0,0 +1,372 @@
+                    module => new Dictionary<string, string>
+                    {
+                        [KubernetesConstants.K8sEdgeModuleLabel] = moduleIdentities[module.Name].DeploymentName(),
+                        [KubernetesConstants.K8sEdgeDeviceLabel] = KubeUtils.SanitizeLabelValue(this.resourceName.DeviceId),
+                        [KubernetesConstants.K8sEdgeHubNameLabel] = KubeUtils.SanitizeLabelValue(this.resourceName.Hostname)
+                    });
+
+            var desiredServices = modules
+                .Select(module => this.serviceProvider.GetService(moduleIdentities[module.Name], module, labels[module.Name]))
+                .Where(service => service.HasValue)
+                .Select(service => service.OrDefault())
+                .ToList();
+
+            V1ServiceList currentServices = await this.client.ListNamespacedServiceAsync(this.deviceNamespace, labelSelector: this.deploymentSelector);
+            await this.ManageServices(currentServices, desiredServices);
+
+            var desiredDeployments = modules
+                .Select(module => this.deploymentProvider.GetDeployment(moduleIdentities[module.Name], module, labels[module.Name]))
+                .ToList();
+
+            V1DeploymentList currentDeployments = await this.client.ListNamespacedDeploymentAsync(this.deviceNamespace, labelSelector: this.deploymentSelector);
+            await this.ManageDeployments(currentDeployments, desiredDeployments);
+
+            var desiredServiceAccounts = modules
+                .Select(module => this.serviceAccountProvider.GetServiceAccount(moduleIdentities[module.Name], module, labels[module.Name]))
+                .ToList();
+
+            V1ServiceAccountList currentServiceAccounts = await this.client.ListNamespacedServiceAccountAsync(this.deviceNamespace, labelSelector: this.deploymentSelector);
+            await this.ManageServiceAccounts(currentServiceAccounts, desiredServiceAccounts);
+
+            return desiredModules;
+        }
+
+        async Task ManageServices(V1ServiceList existing, IEnumerable<V1Service> desired)
+        {
+            // Compose creation strings from existing items
+            Dictionary<string, string> creationStrings = GetServiceConfig(existing);
+
+            // find difference between desired and existing services
+            var diff = desired.Diff(existing.Items, service => service.Metadata.Name);
+
+            // Update only those services if configurations have not matched
+            var updatingTask = diff.Updated
+                .Where(service => service.Metadata.Annotations[KubernetesConstants.CreationString] != creationStrings[service.Metadata.Name])
+                .Select(
+                    service =>
+                    {
+                        Events.UpdateService(service);
+                        return this.client.ReplaceNamespacedServiceAsync(service, service.Metadata.Name, this.deviceNamespace);
```

Comment:
We haven't had much luck with "ReplaceNamespacedServiceAsync" in the past.

Reply:
Replace operations require `resourceVersion` to set from an existing resource.
In addition Service wants `ClusterIP` to be set.